rfcs/text/0011_persistable_state_service.md

```diff
@@ -0,0 +1,218 @@
+  inject: (state: P, references: SavedObjectReference[]) => P,
+  // extract function receives state and should return state with references extracted and array of references
+  // default returns same state with empty reference array
+  extract: (state: P) => { state: P, references: SavedObjectReference[] }
+}
+
+class PersistableStatePlugin {
+  setup(core, plugins) {
+    return {
+      // if one of the functions is not provided default is used 
+      // so we can always assume that all of the functions are available for every persistable state 
+      register: (id: string, definition: Partial<PersistableStateDefinition>) => void
+    }
+  }
+  start(core, plugins) {
+    return {
+      // get will always return. If requested persistable state is not defined it will return default set of functions.
+      get: (id: string) => PersistableStateDefinition,
+      // takes the state, extracts the references and returns them + version string
+      beforeSave: (id: string, state: PersistableState) => [PersistableState, SavedObjectReference[], string],
+      // takes the state, references and version and returns latest (migrated) state with references injected
+      afterLoad: (id: string, state: PersistableState, references: SavedObjectReference[], version: string) => PersistableState,
+    }
+  }
+}
+```
+
+
+# Consuming and edge cases
+
+When plugin A wants to store some state that it does not own it should check with persistableStateRegistry for a registered persistable state definition and execute `beforeSave` function.
+
+```ts
+  const stateForSaving = persistableStatePlugin.beforeSave(id, myState);
+
+``` 
+
+WARNING: If state id is known at compile time we should rather import the correct utilities directly from expression plugin. As using the registry can hide dependencies between plugins when possible plugins should directly depend on plugins providing the state they want to store.
+
+## Handling corrupt state
+
+In current proposal handling of corrupt state is up to the implementator of PersistableStateDefinition. State incoming to migrate function could not match with version string provided so author should validate the incoming state first.
+
+## EnhacedDrilldownEmbeddableInput
+
+Drilldown  plugin adds to the state of embeddable:
+
+```ts
+interface EnhancedDrilldownEmbeddableInput extends EmbeddableInput {
+  enhancements: {
+    drillDowns: { events?: Events[];  }
+  }
+}
+```
+Embeddable is aware that it has an enhancements property on its state, where every key represents state of another plugin.
+
+```ts
+const inject = (state, references) => {
+  Object.getKeys(state.enhancements).forEach(key => {
+    State.enhancements[key] = persistableStatePlugin.get(key).inject(state.enhancements[key], references);
+  }
+}
+```
+
+And drilldown plugin registers its state with persistableStateService.
+
+##
+
+## Use a `{pluginId}+{stateType}` pattern for your ID
+
+To avoid clashes with other unknown plugins.
```

Comment:
> Put some examples to show what you mean by stateType?

++

> Enhancer knows to register their enhancements with drilldowns-{EMBEDDABLE_MIGRATOR} and embeddable knows to look up state.enhancements[key]-{EMBEDDABLE_MIGRATOR}.

Do we want to recommend that registrants export these as consts or in an enum or something? This is implied in the examples but not explicitly stated.

Reply:
i am not sure if the registrator and enhancer need to know about what id to use ?

for example lets take a look into drilldowns:
-  `drilldowns` plugin registers a `drilldowns-drilldownState` persistable state definition
- `3rdparty` plugin registers a `3rdparty-drilldownState` persistable state definition

they both add to `dashboardState.enhancements[key]` ... when dashboard is doing migration it simply checks persistable state service for the `key` which for our drilldowns will be `drilldowns-drildownState` and for 3rd party drilldown it will be `3rdparty-drilldownState` .